plutus-core/testlib/PlutusCore/Generators/PIR/GenTm.hs

```diff
@@ -0,0 +1,317 @@
+genFreshNames ss = do
+  used <- getUniques
+  let i = fromEnum $ Set.findMax $ Set.insert (Unique 0) used
+      js = [ j | j <- [1..i], not $ Unique j `Set.member` used ]
+      is = js ++ take (length ss + 10) [i+1..]
+  is' <- liftGen $ shuffle is
+  return [Name (fromString $ s ++ show j) (toEnum j) | (s, j) <- zip ss is']
+
+-- | See `genFreshName`
+genFreshTyName :: String -> GenTm TyName
+genFreshTyName s = TyName <$> genFreshName s
+
+-- | See `genFreshNames`
+genFreshTyNames :: [String] -> GenTm [TyName]
+genFreshTyNames ss = map TyName <$> genFreshNames ss
+
+-- | Generate a name that overlaps with existing names on purpose. If there
+-- are no existing names, generate a fresh name.
+genNotFreshName :: String -> GenTm Name
+genNotFreshName s = do
+  used <- Set.toList <$> getUniques
+  case used of
+    [] -> genFreshName s
+    _  -> liftGen $ elements [ Name (fromString $ s ++ show (unUnique i)) i | i <- used ]
+
+-- | Generate a fresh name most (a bit more than 75%) of the time and otherwise
+-- generate an already bound name. When there are no bound names generate a fresh name.
+genMaybeFreshName :: String -> GenTm Name
+genMaybeFreshName s = frequencyTm [(3, genFreshName s), (1, genNotFreshName s)]
+
+-- | See `genMaybeFreshName`
+genMaybeFreshTyName :: String -> GenTm TyName
+genMaybeFreshTyName s = TyName <$> genMaybeFreshName s
+
+-- | Bind a type name to a kind and avoid capturing free type variables.
+bindTyName :: TyName -> Kind () -> GenTm a -> GenTm a
+bindTyName x k = local $ \ e -> e { geTypes = Map.insert x k (geTypes e)
+                                  , geTerms = Map.filter (\ty -> not $ x `Set.member` ftvTy ty) (geTerms e)
+                                  , geDatas = Map.delete x (geDatas e)
+                                  }
+
+-- | Bind type names
+bindTyNames :: [(TyName, Kind ())] -> GenTm a -> GenTm a
+bindTyNames = flip $ foldr (uncurry bindTyName)
+
+-- | Remember that we have generated a type name locally but don't bind it.
+-- Useful for non-recursive definitions where we want to control name overlap.
+registerTyName :: TyName -> GenTm a -> GenTm a
+registerTyName n = local $ \ e -> e { geUnboundUsedTyNames = Set.insert n (geUnboundUsedTyNames e) }
+
+-- | Bind a term to a type in a generator.
+bindTmName :: Name -> Type TyName DefaultUni () -> GenTm a -> GenTm a
+bindTmName x ty = local $ \ e -> e { geTerms = Map.insert x ty (geTerms e) }
+
+-- | Bind term names
+bindTmNames :: [(Name, Type TyName DefaultUni ())] -> GenTm a -> GenTm a
+bindTmNames = flip $ foldr (uncurry bindTmName)
+
+-- | Create a fresh term name, bind it to a type, and use it in a generator.
+bindFreshTmName :: String -> Type TyName DefaultUni () -> (Name -> GenTm a) -> GenTm a
+bindFreshTmName name ty k = do
+  x <- genFreshName name
+  bindTmName x ty (k x)
+
+-- Containers (zipper-ish, very useful for shrinking.)
+-- CODE REVIEW: should these go elsewhere? Do these already exist somewhere?
```

Comment:
No idea, but it seems like it should exist...

Reply:
Yeah, I feel like it but I haven't seen this precise version anywhere before.